music/src/core/recorder.ts

```diff
@@ -0,0 +1,168 @@
+/**
+ * A module containing a MIDI recorder.
+ *
+ * @license
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {NoteSequence} from '../protobuf';
+
+/**
+ * An abstract base class for providing arbitrary callbacks for each note
+ * recorded.
+ */
+export abstract class BaseRecorderCallback {
+  /**
+   * Will be called for each time a note is recorded.
+   *
+   * @param seq The note sequence up to this point
+   */
+  abstract run(seq: NoteSequence): void;
+}
+
+/**
+ * Class that records MIDI from any MIDI connected instrument, and converts it
+ * to a `NoteSequence`.
+ */
+export class Recorder {
+  private recording: boolean;
+  public callbackObject: BaseRecorderCallback;
+  private firstNoteTimestamp: number;
+  private notes: NoteSequence.Note[] = [];
+  private downNotes: {[pitch: number]: NoteSequence.Note} = {};
```

Comment:
onNotes I think is better -- also how about using Map<number, NoteSequence.Note>?

Reply:
oh yeah that's definitely what I was trying to say